src/coroutine/memoset/mod.rs

```diff
@@ -619,9 +630,14 @@ impl<F: LurkField, Q: Query<F>> Scope<Q, LogMemo<F>, F> {
 
             if let Some(dependents) = self.dependents.get(query) {
                 for dependent in dependents {
-                    if provenances.get(dependent).is_none() {
-                        next.insert(dependent);
-                    }
+                    missing_dependency_counts
+                        .entry(dependent)
+                        .and_modify(|missing_count| {
+                            *missing_count -= 1;
+                            if *missing_count == 0 {
+                                next.insert(dependent);
+                            }
+                        });
```

Comment:
... which means IIUC that here we may receive a `missing_dependency_counts` with some `(key, 0)` entry.
This would then [underflow](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=628e8383dca9f111f76fcc5af85fc141)
```rust
                        .and_modify(|missing_count| {
                            *missing_count -= 1;
                            if *missing_count == 0 {
                                next.insert(dependent);
                            }
                        }
```

Is this the behavior we want?

Reply:
No, we are looking up the missing dependency counts for the *dependent* of the query that is ready. In other words, after we process this query, then everything that depends on it is a little more ready.

Whenever `missing_count` actually reaches zero, that query becomes immediately ready. The situation you are worried about could (maybe?) happen only if a query depends on itself. However, this is forbidden and already leads to unprovability. Moreover, it can't *actually* happen because (besides being illegal and unprovable via the very provenances we are assembling here) such a query would never have become ready in the first place.

It's possible there is some other bug in this code, but I don't think your concern here represents a problem. The whole point is that we are constructing the provenances bottom-up as we topologically sort. This process also implies that each query will only ever be have its `missing_count` decremented *before* it is processed. Once that count reaches 0, it will never be affected again, because all the queries on which it depends have already been processed previously.

Reply:
This does point out that there is a way for a query to not have its provenance created. Any cyclic computation will never become ready. This is good, since we would never be able to generate a proof. Therefore, we could add an extra check after computing all provenances. If any provenances are missing then we have detected a cyclic computation and can eagerly abort for clear reason. cc: @gabriel-barrett

Reply:
In other words, the impossibility of generating a provenance for a cyclic computation is *exactly* the point we force the prover to manifest provenance in the first place. The benefit of the bottom-up topological sorting vs top-down is that it lets us detect the error condition explicitly, rather than enter into an infinite loop or wild-goose-chase that exhausts the stack.

Reply:
I added the check here: https://github.com/lurk-lab/lurk-rs/pull/1138/commits/3f30a852486767254f6719189e2e3ea1aaf0c9ae. If an application manages to allow for and actually try to prove a cyclic computation, it will be caught. If this happens, the `Lang` supplying a `Query` enabling it is noncompliant.

Reply:
What you're saying is you're unlikely to look up the query that would here provoke an underflow, especially since you end up adding an additional check for cyclic dependencies. In the current form of the code, that's probably right.

In order to prevent a future situation where an errant developer might edit this code without understanding all its implications, simply replacing `*missing_count -=1` by `*missing_count = missing_count.checked_sub(1).expect("ready queries should not be looked up at this stage")`[^1] could provide the local context you just offered, and avoids us having to look into a query never being processed because its `missing_count` has just become enormous.

[^1]: or some more enlightening message

Reply:
Done.